components/script/dom/notification.rs

```diff
@@ -0,0 +1,545 @@
+    /// <https://notifications.spec.whatwg.org/#concept-title>
+    title: DOMString,
+    /// <https://notifications.spec.whatwg.org/#body>
+    body: DOMString,
+    /// <https://notifications.spec.whatwg.org/#data>
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
+    /// <https://notifications.spec.whatwg.org/#concept-direction>
+    dir: NotificationDirection,
+    /// <https://notifications.spec.whatwg.org/#image-url>
+    image: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#icon-url>
+    icon: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#badge-url>
+    badge: Option<USVString>,
+    /// <https://notifications.spec.whatwg.org/#concept-language>
+    lang: DOMString,
+    /// <https://notifications.spec.whatwg.org/#silent-preference-flag>
+    silent: Option<bool>,
+    /// <https://notifications.spec.whatwg.org/#tag>
+    tag: DOMString,
+    /// <https://notifications.spec.whatwg.org/#concept-origin>
+    #[no_trace] // ImmutableOrigin is not traceable
+    origin: ImmutableOrigin,
+    /// <https://notifications.spec.whatwg.org/#vibration-pattern>
+    vibration_pattern: Vec<u32>,
+    /// <https://notifications.spec.whatwg.org/#timestamp>
+    timestamp: u64,
+    /// <https://notifications.spec.whatwg.org/#renotify-preference-flag>
+    renotify: bool,
+    /// <https://notifications.spec.whatwg.org/#require-interaction-preference-flag>
+    require_interaction: bool,
+    /// <https://notifications.spec.whatwg.org/#actions>
+    actions: Vec<Action>,
+    // TODO: image resource, icon resource, and badge resource
+}
+
+impl Notification {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn new(
+        global: &GlobalScope,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: ImmutableOrigin,
+        base_url: ServoUrl,
+        fallback_timestamp: u64,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+    ) -> DomRoot<Self> {
+        reflect_dom_object_with_proto(
+            Box::new(Notification::new_inherited(
+                global,
+                title,
+                options,
+                origin,
+                base_url,
+                fallback_timestamp,
+            )),
+            global,
+            proto,
+            can_gc,
+        )
+    }
+
+    /// partial implementation of <https://notifications.spec.whatwg.org/#create-a-notification>
+    fn new_inherited(
+        global: &GlobalScope,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: ImmutableOrigin,
+        base_url: ServoUrl,
+        fallback_timestamp: u64,
+    ) -> Self {
+        // TODO: missing call to https://html.spec.whatwg.org/multipage/#structuredserializeforstorage
+        // may be find in `dom/bindings/structuredclone.rs`
+        let data = Heap::boxed(options.data.get());
```

Comment:
With my suggestion to remove the box, this can be a `Heap::default()`, then call `.set(...)` on the new field in `Notification::new` after `reflect_dom_object_with_proto`.